File: 0046_permutations.py
# https://leetcode-cn.com/problems/permutations/
class Solution:
    def permute(self, nums):
        """
        Inputs:
            nums: list[int]
        Outputs:
            res: list[list[int]]
        """

        if len(nums) == 1:
            return [nums]
        else:
            val = nums.pop()
            res_tmp = self.permute(nums)
            res = []
            for i in range(len(res_tmp)):
                list_o = res_tmp[i] # the original list

                # Insert val at different locations.
                for j in range(len(list_o)+1):
                    list_c = list_o.copy()
                    list_c.insert(j, val)
                    res.append(list_c)

            return res

File: Cpp/0046_permutations.cpp
// https://leetcode.cn/problems/jump-game-ii/

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res = {{}};
        for (int num: nums) {
            int n = res.size();
            for (int i = 0; i < n; ++i) {
                vector<int> vec = res[i];
                for (int j = 0; j <= vec.size(); ++j) {
                    vector<int> res_tmp(vec);
                    res_tmp.insert(res_tmp.begin()+j, num);
                    res.push_back(res_tmp);
                }
            }
            res.erase(res.begin(), res.begin()+n);
        }
        return res;
    }
};
